File: android/app/src/main/java/com/ooplab/exercises_fitfuel/PoseOverlayView.kt
package com.ooplab.exercises_fitfuel

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.DashPathEffect
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.View
import kotlin.math.*

class PoseOverlayView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null
) : View(context, attrs) {

    enum class HeightGuideState { HIDDEN, TOO_HIGH, TOO_LOW, OK }

    private val TARGET_GUIDE_Y = 0.45f

    private val dotPaint = Paint().apply {
        color = Color.GREEN; style = Paint.Style.FILL; isAntiAlias = true
    }
    private val linePaint = Paint().apply {
        color = Color.WHITE; style = Paint.Style.STROKE; strokeWidth = 6f; isAntiAlias = true
    }
    private val estimatedDotPaint = Paint().apply {
        color = Color.argb(160, 0, 220, 0); style = Paint.Style.FILL; isAntiAlias = true
    }
    private val estimatedLinePaint = Paint().apply {
        color = Color.WHITE; style = Paint.Style.STROKE; strokeWidth = 4f; isAntiAlias = true
        pathEffect = DashPathEffect(floatArrayOf(14f, 9f), 0f)
    }
    private val guideLinePaint = Paint().apply {
        style = Paint.Style.STROKE; strokeWidth = 3f; isAntiAlias = true
        pathEffect = DashPathEffect(floatArrayOf(22f, 13f), 0f)
    }
    private val guideTickPaint  = Paint().apply { style = Paint.Style.STROKE; strokeWidth = 5f; isAntiAlias = true }
    private val guideLabelPaint = Paint().apply { textSize = 42f; isAntiAlias = true; typeface = Typeface.DEFAULT_BOLD }

    // ROSA angle arc paints — instance fields so no allocation per frame
    private val arcPaint = Paint().apply {
        color = Color.parseColor("#FF5722"); style = Paint.Style.STROKE; strokeWidth = 4f; isAntiAlias = true
    }
    private val vertRefPaint = Paint().apply {
        color = Color.argb(200, 255, 87, 34); style = Paint.Style.STROKE; strokeWidth = 3f; isAntiAlias = true
        pathEffect = DashPathEffect(floatArrayOf(12f, 8f), 0f)
    }
    private val angleLabelPaint = Paint().apply {
        color = Color.parseColor("#FF5722"); textSize = 36f; isAntiAlias = true
        typeface = Typeface.DEFAULT_BOLD; textAlign = Paint.Align.CENTER
    }

    private var landmarks: List<LandmarkPoint> = emptyList()
    private var imageWidth: Int = 1
    private var imageHeight: Int = 1
    private var heightGuideState = HeightGuideState.HIDDEN
    private var rosaAngles: RosaAnglesCalculator.Angles? = null
    // Landmark indices to omit from both connections and dots. Empty for the side
    // shots (full skeleton); set during the front phase to mirror the captured
    // photo (face/ears, wrist/palm and legs dropped).
    private var excludedIndices: Set<Int> = emptySet()
    // Front-view only: vertical reference line through each shoulder + the
    // shoulder→elbow angle.
    private var showShoulderAngles = false

    fun updateLandmarks(newLandmarks: List<LandmarkPoint>, imgWidth: Int, imgHeight: Int) {
        landmarks = newLandmarks
        imageWidth = imgWidth
        imageHeight = imgHeight
        postInvalidate()
    }

    fun setExcludedIndices(indices: Set<Int>) {
        if (excludedIndices == indices) return
        excludedIndices = indices
        postInvalidate()
    }

    fun setShowShoulderAngles(show: Boolean) {
        if (showShoulderAngles == show) return
        showShoulderAngles = show
        postInvalidate()
    }

    fun updateRosaAngles(angles: RosaAnglesCalculator.Angles?) {
        rosaAngles = angles
        postInvalidate()
    }

    fun setHeightGuide(state: HeightGuideState) {
        if (heightGuideState == state) return
        heightGuideState = state
        postInvalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (imageWidth <= 1) return

        val scale   = minOf(width.toFloat() / imageWidth, height.toFloat() / imageHeight)
        val offsetX = (width  - imageWidth  * scale) / 2f
        val offsetY = (height - imageHeight * scale) / 2f

        fun sx(x: Float) = x * imageWidth  * scale + offsetX
        fun sy(y: Float) = y * imageHeight * scale + offsetY

        // ── Height guide line ──────────────────────────────────────────────────
        if (heightGuideState != HeightGuideState.HIDDEN) {
            val lineY = sy(TARGET_GUIDE_Y)
            val baseColor = if (heightGuideState == HeightGuideState.OK) Color.GREEN else Color.RED
            guideLinePaint.color = Color.argb(200, Color.red(baseColor), Color.green(baseColor), Color.blue(baseColor))
            canvas.drawLine(0f, lineY, width.toFloat(), lineY, guideLinePaint)
            guideTickPaint.color = baseColor
            val cx = width / 2f
            canvas.drawLine(cx - 40f, lineY - 18f, cx - 40f, lineY + 18f, guideTickPaint)
            canvas.drawLine(cx + 40f, lineY - 18f, cx + 40f, lineY + 18f, guideTickPaint)
            canvas.drawLine(cx - 40f, lineY, cx - 14f, lineY, guideTickPaint)
            canvas.drawLine(cx + 14f, lineY, cx + 40f, lineY, guideTickPaint)
            val label = when (heightGuideState) {
                HeightGuideState.TOO_HIGH -> "▼  Lower phone"
                HeightGuideState.TOO_LOW  -> "▲  Raise phone"
                HeightGuideState.OK       -> "✓  Height OK"
                HeightGuideState.HIDDEN   -> ""
            }
            guideLabelPaint.color = baseColor
            canvas.drawText(label, cx + 56f, lineY - 12f, guideLabelPaint)
        }

        if (landmarks.isEmpty()) return

        // ── Skeleton connections ───────────────────────────────────────────────
        for ((start, end) in POSE_CONNECTIONS) {
            if (start in excludedIndices || end in excludedIndices) continue
            if (start < landmarks.size && end < landmarks.size) {
                val s = landmarks[start]; val e = landmarks[end]
                val paint = if (s.estimated || e.estimated) estimatedLinePaint else linePaint
                canvas.drawLine(sx(s.x), sy(s.y), sx(e.x), sy(e.y), paint)
            }
        }

        // ── Landmark dots ──────────────────────────────────────────────────────
        for ((i, lm) in landmarks.withIndex()) {
            if (i in excludedIndices) continue
            if (lm.estimated) canvas.drawCircle(sx(lm.x), sy(lm.y), 5f, estimatedDotPaint)
            else              canvas.drawCircle(sx(lm.x), sy(lm.y), 7f, dotPaint)
        }

        // ── ROSA angle arcs ────────────────────────────────────────────────────
        val ra = rosaAngles
        if (ra != null && landmarks.size >= 29) {
            drawAngles(canvas, landmarks, ra, ::sx, ::sy, arcPaint, vertRefPaint, angleLabelPaint)
        }

        // ── Front-view shoulder verticals + angle ────────────────────────────────
        if (showShoulderAngles) {
            drawShoulderVerticals(canvas, landmarks, ::sx, ::sy, vertRefPaint, arcPaint, angleLabelPaint)
        }
    }

    companion object {
        val POSE_CONNECTIONS = listOf(
            0 to 1, 1 to 2, 2 to 3, 3 to 7,
            0 to 4, 4 to 5, 5 to 6, 6 to 8,
            9 to 10,
            7 to 11, 8 to 12,
            11 to 12,
            11 to 13, 13 to 15, 15 to 17, 17 to 19, 19 to 15, 15 to 21,
            12 to 14, 14 to 16, 16 to 18, 18 to 20, 20 to 16, 16 to 22,
            11 to 23, 12 to 24, 23 to 24,
            23 to 25, 25 to 27, 27 to 29, 29 to 31, 31 to 27,
            24 to 26, 26 to 28, 28 to 30, 30 to 32, 32 to 28
        )

        /** MediaPipe's canonical 21-point hand skeleton: thumb, four fingers, palm. */
        val HAND_CONNECTIONS = listOf(
            0 to 1, 1 to 2, 2 to 3, 3 to 4,        // thumb
            0 to 5, 5 to 6, 6 to 7, 7 to 8,        // index
            5 to 9, 9 to 10, 10 to 11, 11 to 12,   // middle
            9 to 13, 13 to 14, 14 to 15, 15 to 16, // ring
            13 to 17, 17 to 18, 18 to 19, 19 to 20,// pinky
            0 to 17,                                // palm base
        )

        /** Draws the detected hands and the elbow→wrist connector. Shared by the live
         *  overlay and bakeSkeletonOntoPhoto so both look identical. Uses the same
         *  [linePaint]/[dotPaint] as the body skeleton so the hand reads as a
         *  continuation of the arm. [poseLandmarks] supplies the elbows (13/14);
         *  each hand's wrist (point 0) is joined to its nearest elbow. */
        fun drawHands(
            canvas: Canvas,
            hands: List<List<LandmarkPoint>>,
            poseLandmarks: List<LandmarkPoint>,
            sx: (Float) -> Float,
            sy: (Float) -> Float,
            linePaint: Paint,
            dotPaint: Paint,
            dotRadius: Float,
            arcPaint: Paint,
            labelPaint: Paint,
        ) {
            val elbows = listOfNotNull(poseLandmarks.getOrNull(13), poseLandmarks.getOrNull(14))
            for (hand in hands) {
                val h0 = hand.getOrNull(0)
                // Nearest elbow by squared pixel distance (comparable, no sqrt).
                val nearest = if (h0 != null && elbows.isNotEmpty()) elbows.minByOrNull {
                    val dx = sx(it.x) - sx(h0.x); val dy = sy(it.y) - sy(h0.y)
                    dx * dx + dy * dy
                } else null
                if (h0 != null && nearest != null) {
                    canvas.drawLine(sx(nearest.x), sy(nearest.y), sx(h0.x), sy(h0.y), linePaint)
                }
                for ((start, end) in HAND_CONNECTIONS) {
                    if (start < hand.size && end < hand.size) {
                        canvas.drawLine(sx(hand[start].x), sy(hand[start].y),
                                        sx(hand[end].x), sy(hand[end].y), linePaint)
                    }
                }
                // Dotted axis: wrist (0) → middle-finger MCP (9).
                if (hand.size > 9) {
                    val sw = linePaint.strokeWidth
                    val dashed = Paint(linePaint).apply {
                        pathEffect = DashPathEffect(floatArrayOf(sw * 2f, sw * 1.4f), 0f)
                    }
                    canvas.drawLine(sx(hand[0].x), sy(hand[0].y),
                                    sx(hand[9].x), sy(hand[9].y), dashed)
                }
                // Wrist angle: forearm (elbow→wrist) vs hand axis (wrist→MCP).
                // ~180° when the hand is straight in line with the forearm.
                if (h0 != null && nearest != null && hand.size > 9) {
                    val mcp = hand[9]
                    val v1x = (sx(nearest.x) - sx(h0.x)).toDouble(); val v1y = (sy(nearest.y) - sy(h0.y)).toDouble()
                    val v2x = (sx(mcp.x) - sx(h0.x)).toDouble();     val v2y = (sy(mcp.y) - sy(h0.y)).toDouble()
                    val m1 = hypot(v1x, v1y); val m2 = hypot(v2x, v2y)
                    if (m1 >= 1.0 && m2 >= 1.0) {
                        val cosA = ((v1x * v2x + v1y * v2y) / (m1 * m2)).coerceIn(-1.0, 1.0)
                        val angle = Math.toDegrees(acos(cosA))
                        drawArc(canvas, sx(nearest.x), sy(nearest.y), sx(h0.x), sy(h0.y),
                            sx(mcp.x), sy(mcp.y), "${angle.roundToInt()}°",
                            (m2 * 0.6).toFloat(), arcPaint, labelPaint)
                    }
                }
                for (lm in hand) canvas.drawCircle(sx(lm.x), sy(lm.y), dotRadius, dotPaint)
            }
        }

        /** Draws a vertical dotted reference line through each shoulder (11/12) and
         *  labels the angle between that vertical and the shoulder→elbow line
         *  (11→13, 12→14). Front-view only. Shared by the live overlay and the baked
         *  photo so both look identical. */
        fun drawShoulderVerticals(
            canvas: Canvas,
            landmarks: List<LandmarkPoint>,
            sx: (Float) -> Float,
            sy: (Float) -> Float,
            vertPaint: Paint,
            arcPaint: Paint,
            labelPaint: Paint,
        ) {
            for ((shIdx, elIdx) in listOf(11 to 13, 12 to 14)) {
                if (shIdx >= landmarks.size || elIdx >= landmarks.size) continue
                val shX = sx(landmarks[shIdx].x); val shY = sy(landmarks[shIdx].y)
                val elX = sx(landmarks[elIdx].x); val elY = sy(landmarks[elIdx].y)
                val dx = elX - shX; val dy = elY - shY
                val len = hypot(dx.toDouble(), dy.toDouble()).toFloat()
                if (len < 1f) continue

                // Vertical dotted line through the shoulder.
                canvas.drawLine(shX, shY - 0.4f * len, shX, shY + len, vertPaint)

                // Angle arc between the shoulder→elbow line and the straight-down
                // vertical — same arc + label style as the side-view ROSA angles.
                val angleDeg = Math.toDegrees(acos((dy / len).toDouble())).toFloat()
                drawArc(canvas, elX, elY, shX, shY, shX, shY + len,
                    "${angleDeg.roundToInt()}°", len * 0.3f, arcPaint, labelPaint)
            }
        }

        /** Draws the three ROSA angle arcs onto [canvas]. Called from both the live
         *  overlay (onDraw) and bakeSkeletonOntoPhoto, so both views look identical. */
        fun drawAngles(
            canvas: Canvas,
            landmarks: List<LandmarkPoint>,
            angles: RosaAnglesCalculator.Angles,
            sx: (Float) -> Float,
            sy: (Float) -> Float,
            arcPaint: Paint,
            vertRefPaint: Paint,
            labelPaint: Paint,
        ) {
            if (landmarks.size < 29) return
            val left = angles.isLeftSide

            val ear      = landmarks[if (left) 7 else 8]
            val shoulder = landmarks[if (left) 11 else 12]
            val elbow    = landmarks[if (left) 13 else 14]
            val wrist    = landmarks[if (left) 15 else 16]
            val hip      = landmarks[if (left) 23 else 24]
            val knee     = landmarks[if (left) 25 else 26]
            val ankle    = landmarks[if (left) 27 else 28]

            val shSx = sx(shoulder.x); val shSy = sy(shoulder.y)
            val hipSx = sx(hip.x);     val hipSy = sy(hip.y)

            // Arc radius proportional to torso height — scales with person's distance
            val torsoLen = hypot((shSx - hipSx).toDouble(), (shSy - hipSy).toDouble()).toFloat()
            val r = torsoLen * 0.22f

            // Knee angle — hip → knee ← ankle
            drawArc(canvas, sx(hip.x), sy(hip.y), sx(knee.x), sy(knee.y),
                sx(ankle.x), sy(ankle.y), "${"%.0f".format(angles.kneeAngle)}°", r, arcPaint, labelPaint)

            // Trunk angle — shoulder → hip ← vertical reference line
            val vRefSy = sy(hip.y - 0.18f)
            canvas.drawLine(hipSx, hipSy, hipSx, vRefSy, vertRefPaint)
            drawArc(canvas, shSx, shSy, hipSx, hipSy, hipSx, vRefSy,
                "${"%.0f".format(angles.trunkAngle)}°", r, arcPaint, labelPaint)

            // Neck angle — ear → shoulder ← vertical reference line
            val neckVRefSy = sy(shoulder.y - 0.14f)
            canvas.drawLine(shSx, shSy, shSx, neckVRefSy, vertRefPaint)
            drawArc(canvas, sx(ear.x), sy(ear.y), shSx, shSy, shSx, neckVRefSy,
                "${"%.0f".format(angles.neckAngle)}°", r * 0.75f, arcPaint, labelPaint)

            // Elbow angle — shoulder → elbow ← wrist
            drawArc(canvas, shSx, shSy, sx(elbow.x), sy(elbow.y),
                sx(wrist.x), sy(wrist.y), "${"%.0f".format(angles.elbowAngle)}°", r, arcPaint, labelPaint)
        }

        private fun drawArc(
            canvas: Canvas,
            p1x: Float, p1y: Float,
            vx: Float,  vy: Float,
            p2x: Float, p2y: Float,
            text: String,
            radius: Float,
            arcPaint: Paint,
            labelPaint: Paint,
        ) {
            val v1x = p1x - vx; val v1y = p1y - vy
            val v2x = p2x - vx; val v2y = p2y - vy
            val m1 = sqrt((v1x * v1x + v1y * v1y).toDouble()).toFloat()
            val m2 = sqrt((v2x * v2x + v2y * v2y).toDouble()).toFloat()
            if (m1 < 1f || m2 < 1f) return

            val startAngle = Math.toDegrees(atan2(v1y.toDouble(), v1x.toDouble())).toFloat()
            val cross      = v1x * v2y - v1y * v2x
            val dot        = v1x * v2x + v1y * v2y
            val sweepAngle = Math.toDegrees(atan2(cross.toDouble(), dot.toDouble())).toFloat()

            val rect = RectF(vx - radius, vy - radius, vx + radius, vy + radius)

            // Black outline pass — makes arc visible on both white and black backgrounds
            val outlineArcPaint = Paint(arcPaint).apply {
                color = Color.BLACK; strokeWidth = arcPaint.strokeWidth + 4f
            }
            canvas.drawArc(rect, startAngle, sweepAngle, false, outlineArcPaint)
            canvas.drawArc(rect, startAngle, sweepAngle, false, arcPaint)

            val bisectorRad = Math.toRadians((startAngle + sweepAngle / 2f).toDouble())
            val textDist = radius + labelPaint.textSize * 0.6f
            val textX = vx + (textDist * cos(bisectorRad)).toFloat()
            val textY = vy + (textDist * sin(bisectorRad)).toFloat() + labelPaint.textSize / 3f

            // Black outline pass for text
            val outlineLabelPaint = Paint(labelPaint).apply {
                color = Color.BLACK; style = Paint.Style.STROKE; strokeWidth = labelPaint.textSize * 0.22f
            }
            canvas.drawText(text, textX, textY, outlineLabelPaint)
            canvas.drawText(text, textX, textY, labelPaint)
        }
    }
}

File: ios/Runner/PostureEngine/PoseOverlayView.swift
import UIKit

/// Live skeleton + ROSA-angle overlay drawn over the camera preview.
/// Port of the Kotlin `PoseOverlayView`. The actual skeleton/angle drawing lives
/// in `PoseRenderer` so the captured-photo baking path can reuse it verbatim
/// (mirroring how Android shares `drawAngles` between the view and bakeSkeleton).
final class PoseOverlayView: UIView {

    enum HeightGuideState { case hidden, tooHigh, tooLow, ok }

    private let TARGET_GUIDE_Y: CGFloat = 0.45

    private var landmarks: [LandmarkPoint] = []
    private var imageWidth: CGFloat = 1
    private var imageHeight: CGFloat = 1
    private var heightGuideState: HeightGuideState = .hidden
    private var rosaAngles: RosaAnglesCalculator.Angles?

    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
        isOpaque = false
        isUserInteractionEnabled = false
    }
    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    func updateLandmarks(_ newLandmarks: [LandmarkPoint], imgWidth: Int, imgHeight: Int) {
        landmarks = newLandmarks
        imageWidth = CGFloat(max(imgWidth, 1))
        imageHeight = CGFloat(max(imgHeight, 1))
        setNeedsDisplay()
    }

    func updateRosaAngles(_ angles: RosaAnglesCalculator.Angles?) {
        rosaAngles = angles
        setNeedsDisplay()
    }

    func setHeightGuide(_ state: HeightGuideState) {
        if heightGuideState == state { return }
        heightGuideState = state
        setNeedsDisplay()
    }

    override func draw(_ rect: CGRect) {
        guard let ctx = UIGraphicsGetCurrentContext(), imageWidth > 1 else { return }

        let scale = min(bounds.width / imageWidth, bounds.height / imageHeight)
        let offsetX = (bounds.width - imageWidth * scale) / 2
        let offsetY = (bounds.height - imageHeight * scale) / 2
        let sx: (Float) -> CGFloat = { CGFloat($0) * self.imageWidth * scale + offsetX }
        let sy: (Float) -> CGFloat = { CGFloat($0) * self.imageHeight * scale + offsetY }

        // ── Height guide line ──────────────────────────────────────────────────
        if heightGuideState != .hidden {
            let lineY = sy(Float(TARGET_GUIDE_Y))
            let baseColor: UIColor = heightGuideState == .ok ? .green : .red
            let lineColor = baseColor.withAlphaComponent(200.0 / 255.0)
            ctx.setStrokeColor(lineColor.cgColor)
            ctx.setLineWidth(3)
            ctx.setLineDash(phase: 0, lengths: [22, 13])
            ctx.move(to: CGPoint(x: 0, y: lineY))
            ctx.addLine(to: CGPoint(x: bounds.width, y: lineY))
            ctx.strokePath()
            ctx.setLineDash(phase: 0, lengths: [])

            ctx.setStrokeColor(baseColor.cgColor)
            ctx.setLineWidth(5)
            let cx = bounds.width / 2
            for seg in [(cx - 40, lineY - 18, cx - 40, lineY + 18),
                        (cx + 40, lineY - 18, cx + 40, lineY + 18),
                        (cx - 40, lineY, cx - 14, lineY),
                        (cx + 14, lineY, cx + 40, lineY)] {
                ctx.move(to: CGPoint(x: seg.0, y: seg.1))
                ctx.addLine(to: CGPoint(x: seg.2, y: seg.3))
            }
            ctx.strokePath()

            let label: String
            switch heightGuideState {
            case .tooHigh: label = "▼  Lower phone"
            case .tooLow: label = "▲  Raise phone"
            case .ok: label = "✓  Height OK"
            case .hidden: label = ""
            }
            let attrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.boldSystemFont(ofSize: 42),
                .foregroundColor: baseColor,
            ]
            (label as NSString).draw(at: CGPoint(x: cx + 56, y: lineY - 12 - 42), withAttributes: attrs)
        }

        if landmarks.isEmpty { return }

        PoseRenderer.drawScene(in: ctx, landmarks: landmarks, angles: rosaAngles, sx: sx, sy: sy, scale: 1)
    }

    static let poseConnections = PoseRenderer.poseConnections
}

/// Stateless skeleton/angle renderer shared by the live overlay and the
/// captured-photo baking step.
enum PoseRenderer {

    static let poseConnections: [(Int, Int)] = [
        (0, 1), (1, 2), (2, 3), (3, 7),
        (0, 4), (4, 5), (5, 6), (6, 8),
        (9, 10),
        (7, 11), (8, 12),
        (11, 12),
        (11, 13), (13, 15), (15, 17), (17, 19), (19, 15), (15, 21),
        (12, 14), (14, 16), (16, 18), (18, 20), (20, 16), (16, 22),
        (11, 23), (12, 24), (23, 24),
        (23, 25), (25, 27), (27, 29), (29, 31), (31, 27),
        (24, 26), (26, 28), (28, 30), (30, 32), (32, 28),
    ]

    private static let arcColor = UIColor(red: 0xFF / 255, green: 0x57 / 255, blue: 0x22 / 255, alpha: 1)
    private static let vertRefColor = UIColor(red: 255 / 255, green: 87 / 255, blue: 34 / 255, alpha: 200 / 255)
    private static let estimatedDotColor = UIColor(red: 0, green: 220 / 255, blue: 0, alpha: 160 / 255)

    /// Draws skeleton connections, landmark dots, and ROSA angle arcs.
    /// `scale` multiplies stroke widths / dot radii / text sizes — 1 for the live
    /// overlay, `photoWidth / referenceWidth` for the baked photo.
    static func drawScene(in ctx: CGContext,
                          landmarks: [LandmarkPoint],
                          angles: RosaAnglesCalculator.Angles?,
                          sx: (Float) -> CGFloat,
                          sy: (Float) -> CGFloat,
                          scale: CGFloat) {
        // ── Skeleton connections ───────────────────────────────────────────────
        for (start, end) in poseConnections where start < landmarks.count && end < landmarks.count {
            let s = landmarks[start], e = landmarks[end]
            let estimated = s.estimated || e.estimated
            ctx.setStrokeColor(UIColor.white.cgColor)
            ctx.setLineWidth((estimated ? 4 : 6) * scale)
            ctx.setLineDash(phase: 0, lengths: estimated ? [14 * scale, 9 * scale] : [])
            ctx.move(to: CGPoint(x: sx(s.x), y: sy(s.y)))
            ctx.addLine(to: CGPoint(x: sx(e.x), y: sy(e.y)))
            ctx.strokePath()
        }
        ctx.setLineDash(phase: 0, lengths: [])

        // ── Landmark dots ──────────────────────────────────────────────────────
        for lm in landmarks {
            let r = (lm.estimated ? 5 : 7) * scale
            ctx.setFillColor((lm.estimated ? estimatedDotColor : UIColor.green).cgColor)
            ctx.fillEllipse(in: CGRect(x: sx(lm.x) - r, y: sy(lm.y) - r, width: 2 * r, height: 2 * r))
        }

        // ── ROSA angle arcs ────────────────────────────────────────────────────
        if let ra = angles, landmarks.count >= 29 {
            drawAngles(in: ctx, landmarks: landmarks, angles: ra, sx: sx, sy: sy, scale: scale)
        }
    }

    static func drawAngles(in ctx: CGContext,
                           landmarks: [LandmarkPoint],
                           angles: RosaAnglesCalculator.Angles,
                           sx: (Float) -> CGFloat,
                           sy: (Float) -> CGFloat,
                           scale: CGFloat) {
        if landmarks.count < 29 { return }
        let left = angles.isLeftSide

        let ear = landmarks[left ? 7 : 8]
        let shoulder = landmarks[left ? 11 : 12]
        let elbow = landmarks[left ? 13 : 14]
        let wrist = landmarks[left ? 15 : 16]
        let hip = landmarks[left ? 23 : 24]
        let knee = landmarks[left ? 25 : 26]
        let ankle = landmarks[left ? 27 : 28]

        let shSx = sx(shoulder.x), shSy = sy(shoulder.y)
        let hipSx = sx(hip.x), hipSy = sy(hip.y)

        let torsoLen = hypot(shSx - hipSx, shSy - hipSy)
        let r = torsoLen * 0.22
        let textSize = 36 * scale

        // Knee angle — hip → knee ← ankle
        drawArc(ctx, p1: CGPoint(x: sx(hip.x), y: sy(hip.y)),
                v: CGPoint(x: sx(knee.x), y: sy(knee.y)),
                p2: CGPoint(x: sx(ankle.x), y: sy(ankle.y)),
                text: deg(angles.kneeAngle), radius: r, scale: scale, textSize: textSize)

        // Trunk angle — shoulder → hip ← vertical reference line
        let vRefSy = sy(hip.y - 0.18)
        strokeVertRef(ctx, from: CGPoint(x: hipSx, y: hipSy), to: CGPoint(x: hipSx, y: vRefSy), scale: scale)
        drawArc(ctx, p1: CGPoint(x: shSx, y: shSy), v: CGPoint(x: hipSx, y: hipSy),
                p2: CGPoint(x: hipSx, y: vRefSy),
                text: deg(angles.trunkAngle), radius: r, scale: scale, textSize: textSize)

        // Neck angle — ear → shoulder ← vertical reference line
        let neckVRefSy = sy(shoulder.y - 0.14)
        strokeVertRef(ctx, from: CGPoint(x: shSx, y: shSy), to: CGPoint(x: shSx, y: neckVRefSy), scale: scale)
        drawArc(ctx, p1: CGPoint(x: sx(ear.x), y: sy(ear.y)), v: CGPoint(x: shSx, y: shSy),
                p2: CGPoint(x: shSx, y: neckVRefSy),
                text: deg(angles.neckAngle), radius: r * 0.75, scale: scale, textSize: textSize)

        // Elbow angle — shoulder → elbow ← wrist
        drawArc(ctx, p1: CGPoint(x: shSx, y: shSy), v: CGPoint(x: sx(elbow.x), y: sy(elbow.y)),
                p2: CGPoint(x: sx(wrist.x), y: sy(wrist.y)),
                text: deg(angles.elbowAngle), radius: r, scale: scale, textSize: textSize)
    }

    private static func deg(_ value: Float) -> String { String(format: "%.0f°", value) }

    private static func strokeVertRef(_ ctx: CGContext, from: CGPoint, to: CGPoint, scale: CGFloat) {
        ctx.setStrokeColor(vertRefColor.cgColor)
        ctx.setLineWidth(3 * scale)
        ctx.setLineDash(phase: 0, lengths: [12 * scale, 8 * scale])
        ctx.move(to: from)
        ctx.addLine(to: to)
        ctx.strokePath()
        ctx.setLineDash(phase: 0, lengths: [])
    }

    private static func drawArc(_ ctx: CGContext,
                                p1: CGPoint, v: CGPoint, p2: CGPoint,
                                text: String, radius: CGFloat, scale: CGFloat, textSize: CGFloat) {
        let v1 = CGPoint(x: p1.x - v.x, y: p1.y - v.y)
        let v2 = CGPoint(x: p2.x - v.x, y: p2.y - v.y)
        let m1 = hypot(v1.x, v1.y), m2 = hypot(v2.x, v2.y)
        if m1 < 1 || m2 < 1 { return }

        let startAngle = atan2(v1.y, v1.x)
        let cross = v1.x * v2.y - v1.y * v2.x
        let dot = v1.x * v2.x + v1.y * v2.y
        let sweep = atan2(cross, dot)

        // Approximate the arc with a polyline so the sweep direction matches Android's
        // drawArc exactly regardless of CG's clockwise convention.
        let steps = max(2, Int(abs(sweep) / (.pi / 60)))
        let arcPath = CGMutablePath()
        for i in 0...steps {
            let a = startAngle + sweep * CGFloat(i) / CGFloat(steps)
            let pt = CGPoint(x: v.x + radius * cos(a), y: v.y + radius * sin(a))
            if i == 0 { arcPath.move(to: pt) } else { arcPath.addLine(to: pt) }
        }

        // Black outline pass, then the coloured arc.
        ctx.addPath(arcPath)
        ctx.setStrokeColor(UIColor.black.cgColor)
        ctx.setLineWidth(4 * scale + 4)
        ctx.strokePath()
        ctx.addPath(arcPath)
        ctx.setStrokeColor(arcColor.cgColor)
        ctx.setLineWidth(4 * scale)
        ctx.strokePath()

        // Label at the arc bisector.
        let bisector = startAngle + sweep / 2
        let textDist = radius + textSize * 0.6
        let textX = v.x + textDist * cos(bisector)
        let textY = v.y + textDist * sin(bisector)
        drawCenteredOutlinedText(ctx, text: text, center: CGPoint(x: textX, y: textY),
                                 textSize: textSize, color: arcColor)
    }

    private static func drawCenteredOutlinedText(_ ctx: CGContext, text: String, center: CGPoint,
                                                 textSize: CGFloat, color: UIColor) {
        let font = UIFont.boldSystemFont(ofSize: textSize)
        let ns = text as NSString
        let outline: [NSAttributedString.Key: Any] = [
            .font: font,
            .strokeColor: UIColor.black,
            .strokeWidth: textSize * 0.22,   // positive = stroke (outline)
        ]
        let fill: [NSAttributedString.Key: Any] = [.font: font, .foregroundColor: color]
        let size = ns.size(withAttributes: fill)
        let origin = CGPoint(x: center.x - size.width / 2, y: center.y - size.height / 2)
        ns.draw(at: origin, withAttributes: outline)
        ns.draw(at: origin, withAttributes: fill)
    }
}
